crates/ledger/src/rewards.rs

```diff
@@ -151,58 +151,55 @@ impl StakeDistribution {
-                            credential,
-                            AccountState {
-                                pool,
-                                lovelace: account.rewards,
-                            },
-                        );
-                    }
-                }
-            }
-        })?;
-
-        db.with_utxo(|rows| {
-            for (_, row) in rows {
-                if let Some(output) = row.borrow() {
-                    if let Some(credential) = output_stake_credential(output) {
-                        let value = output_lovelace(output);
-                        accounts
-                            .entry(credential)
-                            .and_modify(|account| account.lovelace += value);
-                    }
-                }
-            }
-        })?;
-
-        let mut pools: BTreeMap<PoolId, PoolState> = BTreeMap::new();
-        db.with_pools(|rows| {
-            for (pool, row) in rows {
-                if let Some(row) = row.borrow() {
-                    pools.insert(
+    pub fn new(db: &impl Snapshot) -> Result<Self, StoreError> {
+        // TODO: Avoid creating this intermediate map, and directly create the keys/scripts ones.
+        // Then, when looking for _active accounts_ delegated to pools, we can prune those not
+        // referenced anywhere.
+        let mut accounts = db
+            .iter_accounts()?
+            .filter_map(|(credential, account)| {
+                account.delegatee.map(|pool| (pool, credential, account))
+            })
+            .map(|(pool, credential, account)| {
+                (
+                    credential,
+                    AccountState {
                         pool,
-                        PoolState {
-                            stake: 0,
-                            blocks_count: 0,
-                            // NOTE: pre-compute margin here (1 - m), which gets used for all
-                            // member and leader rewards calculation.
-                            margin: safe_ratio(
-                                row.current_params.margin.numerator,
-                                row.current_params.margin.denominator,
-                            ),
-                            parameters: row.current_params.clone(),
-                        },
-                    );
-                }
-            }
-        })?;
+                        lovelace: account.rewards,
+                    },
+                )
+            })
+            .collect::<BTreeMap<StakeCredential, AccountState>>();
+
+        let credentials = db.iter_utxos()?.filter_map(|(_, out)| {
+            output_stake_credential(&out).map(|credential| (out, credential))
+        });
+        credentials.for_each(|(output, credential)| {
+            let value = output_lovelace(&output);
+            accounts
+                .entry(credential)
+                .and_modify(|account| account.lovelace += value);
+        });
```

Comment:
```suggestion
        db.iter_utxos()?.for_each(|(_, output)| {
            if let Some(credential) = output_stake_credential(&output) {
                let value = output_lovelace(&output);
                accounts
                    .entry(credential)
                    .and_modify(|account| account.lovelace += value);
            }
        });
```

So that we only traverse the structure once; there are potentially millions of credentials so it matters quite a lot.